### PR TITLE
Auto level extraction

### PR DESCRIPTION
**Pull request by mkeshavarzi**
Branched from #356. This PR aims to automate the level extraction of the load ifcmodel in runtime through ifclib or utils

### CODE REVIEW

File: src/Components/ExtractLevelsMenu.jsx [hunk 1]
```diff
@@ -105,28 +106,36 @@ export default function ExtractLevelsMenu({listOfOptions, icon, title}) {
...
+    for (let i = 0; i < sampleHeights.length; i++) {
+      sampleHeightsIndex[i] = i
+    }
 
-  const sampleHeightsIndex = [0, 1, 2, 3, 4, 5, 6]
-  /* eslint-enable */
+    const planeoffset = 0.5
+    sampleHeightsIndex.forEach((data) => {
+      floorplanMenu.push(
+          <MenuItem onClick={() => createFloorplanPlane(sampleHeights[data], sampleHeights[data + 1] - planeoffset)}>  L{data} </MenuItem>)
```
Comment (mkeshavarzi) [inline src/Components/ExtractLevelsMenu.jsx:134]: @pablo-mayrgundter, I'm trying to get the generated submenu to show up on the UI after the promise is fulfilled

File: src/Components/ExtractLevelsMenu.jsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  }
+  const planView = () => {
+    // viewer.context.ifcCamera.projectionManager.setOrthoCamera()
+    viewer.context.ifcCamera.toggleProjection()
+    viewer.plans.moveCameraTo2DPlanPosition(true)
+  }
+  const createPlane = (normalDirection) => {
+    const modelCenter = getModelCenter(model)
+    const planeHash = getHashParams(location, 'p')
+    console.log('in the function modelCenter', modelCenter)
```
Comment (OlegMoshkovich) [inline src/Components/ExtractLevelsMenu.jsx:59]: remove.

Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:59]: Remove the entire ``createPlane`` function?

Comment (OlegMoshkovich) [reply src/Components/ExtractLevelsMenu.jsx:59]: There should be no console.logs in production

**Review by OlegMoshkovich — CHANGES_REQUESTED** [review]

File: src/Components/ExtractLevelsMenu.jsx (hunk 2 above)
Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:59]: Already addressed in #356. This commit was branched from #356, and after implementation of the auto-level extraction, #364 and #356 were rebased. Code cleanup took place later in #356.

File: src/Components/ExtractLevelsMenu.jsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+          normal = new Vector3(0, 1, 0)
+          break
+        case 'z':
+          normal = new Vector3(0, 0, 1)
+          break
+        default:
+          normal = new Vector3(0, 1, 0)
+          break
+      }
+      console.log(' ------------------------- ')
```
[the remark this answers is not in the document]
Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:101]: Already addressed in #356. This commit was branched from #356, and after implementation of the auto-level extraction, #364 and #356 were rebased. Code cleanup took place later in #356.

File: src/Components/ExtractLevelsMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        normal = new Vector3(0, 1, 0)
+        break
+    }
+    console.log('in the function normal', normal)
+    if (!planeHash || planeHash !== normalDirection ) {
+      addHashParams(window.location, PLANE_PREFIX, {planeAxis: normalDirection})
+    }
+    return viewer.clipper.createFromNormalAndCoplanarPoint(normal, modelCenter)
+  }
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:81]: Removed.

File: src/utils/extractHeight.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Extract related elements.
+ *
+ * @param {object} ifcModel
+ * @return {Array} elevation values
+ */
+export async function extractHeight(ifcModel) {
+  try {
+    const ifcBuildingStoreyID = 3124254112
```
[the remark this answers is not in the document]
Comment (mkeshavarzi) [reply src/utils/extractHeight.js:9]: Yes. It maps IFCBUILDINGSTOREY. Please see:

https://github.com/IFCjs/web-ifc-three/blob/d133f84181fd10a70ad3e5c0b93b52bf3d330989/web-ifc-three/src/IFC/components/IFCElementsMap.ts#L5

File: src/Components/ExtractLevelsMenu.jsx [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  const handleClick = (event) => {
+    setAnchorEl(event.currentTarget)
+  }
+  const handleClose = () => {
+    setAnchorEl(null)
+  }
+  const viewer = useStore((state) => state.viewerStore)
+  const location = useLocation()
+  const createFloorplanPlane = (h1, h2) => {
+    console.log('Got here')
```
[the remark this answers is not in the document]
Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:42]: Removed in #356.

**Comment by mkeshavarzi** [comment]
Since rebase has happened in #356. Should we just close this PR and continue in #356?

File: src/Components/ExtractLevelsMenu.jsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      console.log('the modelCenter from the useEffect', modelCenter)
+      console.log('plane direction', planeDirection)
+      console.log('normal', normal)
+      console.log(' ------------------------- ')
+      createPlane(planeDirection)
+      // viewer.clipper.createFromNormalAndCoplanarPoint(normal, modelCenter)
+    }
+  }, [model])
+
+  showExtractMenu = async () => {
```
[the remark this answers is not in the document]
Comment (mkeshavarzi) [reply src/Components/ExtractLevelsMenu.jsx:111]: Yes.

File: src/Components/ExtractLevelsMenu.jsx [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      for (let i = 0; i < allStor.length; i++) {
+        floorplanMenuItems[i] = (
+          <MenuItem onClick={() =>
+            createFloorplanPlane(allStor[i], allStor[i + 1] - planeoffset)}
+          >  L{i} </MenuItem>)
+      }
+    }
+  }
+
+  showExtractMenu()
```
[the remark this answers is not in the document]
Comment (OlegMoshkovich) [reply src/Components/ExtractLevelsMenu.jsx:124]: you can use useEffect and make it run based on model changes
https://www.w3schools.com/react/react_useeffect.asp

File: src/Components/ExtractLevelsMenu.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (OlegMoshkovich) [reply src/Components/ExtractLevelsMenu.jsx:111]: This patter does not seem correct, state variables should not be defined directly.
https://reactjs.org/docs/hooks-state.html

File: src/utils/extractHeight.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (OlegMoshkovich) [reply src/utils/extractHeight.js:9]: got it.

**Comment by OlegMoshkovich** [comment]
sure. i don't understand your branch strategy, so do whatever is the easiest for everyone.

File: src/Components/ExtractLevelsMenu.jsx [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    removePlanes(viewer)
+    const modelCenter1 = new Vector3(0, h1, 0)
+    const modelCenter2 = new Vector3(0, h2, 0)
+    const normal1 = new Vector3(0, 1, 0)
+    const normal2 = new Vector3(0, -1, 0)
+    viewer.clipper.createFromNormalAndCoplanarPoint(normal1, modelCenter1)
+    viewer.clipper.createFromNormalAndCoplanarPoint(normal2, modelCenter2)
+  }
+  const planView = () => {
+    // viewer.context.ifcCamera.projectionManager.setOrthoCamera()
```
Comment (OlegMoshkovich) [inline src/Components/ExtractLevelsMenu.jsx:52]: delete

File: src/Components/ExtractLevelsMenu.jsx [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+export default function ExtractLevelsMenu({listOfOptions, icon, title}) {
+  const [anchorEl, setAnchorEl] = useState(null)
+  const classes = useStyles()
+  const open = Boolean(anchorEl)
+
+  const model = useStore((state) => state.modelStore)
+  let [floorplanMenuItems, showExtractMenu] = useState([])
+
+
+  const PLANE_PREFIX = 'p'
```
Comment (OlegMoshkovich) [inline src/Components/ExtractLevelsMenu.jsx:32]: delete

File: src/Components/ExtractLevelsMenu.jsx [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        normal = new Vector3(0, 1, 0)
+        break
+      case 'z':
+        normal = new Vector3(0, 0, 1)
+        break
+      default:
+        normal = new Vector3(0, 1, 0)
+        break
+    }
+    console.log('in the function normal', normal)
```
Comment (OlegMoshkovich) [inline src/Components/ExtractLevelsMenu.jsx:75]: delete

**Pull request merged**